commons/src/env_command.rs

```diff
@@ -0,0 +1,537 @@
+    ///                   .unwrap();
+    ///
+    /// assert!(!outcome.status.success());
+    /// assert!(outcome.stderr.contains("command not found: iDoNotExist"))
+    /// ```
+    pub fn on_non_zero_exit(
+        &mut self,
+        fun: impl Fn(NonZeroExitStatusError) -> Result<EnvCommandResult, NonZeroExitStatusError>
+            + 'static,
+    ) -> &mut Self {
+        self.on_non_zero_exit = Box::new(fun);
+        self
+    }
+
+    /// Internal helper for streaming a child process to stdout/stderr
+    /// while also collecting the results of the process.
+    fn stream_child(child: &mut std::process::Child) -> EnvCommandResult {
+        let child_stdout = child
+            .stdout
+            .take()
+            .expect("Internal error, could not take stdout");
+        let child_stderr = child
+            .stderr
+            .take()
+            .expect("Internal error, could not take stderr");
+
+        let (stdout_tx, stdout_rx) = std::sync::mpsc::channel();
+        let (stderr_tx, stderr_rx) = std::sync::mpsc::channel();
+
+        let stdout_thread = thread::spawn(move || {
+            let stdout_lines = BufReader::new(child_stdout).lines();
+            for line in stdout_lines {
+                let line = line.unwrap();
+                println!("{}", line);
+                stdout_tx.send(line).unwrap();
+            }
+        });
+
+        let stderr_thread = thread::spawn(move || {
+            let stderr_lines = BufReader::new(child_stderr).lines();
+            for line in stderr_lines {
+                let line = line.unwrap();
+                eprintln!("{}", line);
+                stderr_tx.send(line).unwrap();
+            }
+        });
+
+        let status = child
+            .wait()
+            .expect("Internal error, failed to wait on child");
+
+        stdout_thread.join().unwrap();
+        stderr_thread.join().unwrap();
+
+        let stdout = stdout_rx.into_iter().collect::<String>();
+        let stderr = stderr_rx.into_iter().collect::<String>();
+
+        EnvCommandResult {
+            stdout,
+            stderr,
+            status,
+        }
+    }
+
+    /// The user can specify an custom function to be called when
+    /// a non-zero exit status is encountered.
+    ///
+    /// This is a helper function to invoke that behavior.
+    fn handle_non_zero_exit_error(
+        &self,
+        error: NonZeroExitStatusError,
+    ) -> Result<EnvCommandResult, EnvCommandError> {
+        (self.on_non_zero_exit)(error).map_err(EnvCommandError::UnexpectedExitStatusError)
+    }
+
+    pub fn to_string<T: Borrow<HashMap<OsString, OsString>>>(
+        command: &Command,
+        show_env_keys: impl Iterator<Item = impl Into<OsString>>,
+        env: T,
+    ) -> String {
+        let escape_pattern = Regex::new(r"([^A-Za-z0-9_\-.,:/@\n])").unwrap(); // https://github.com/jimmycuadra/rust-shellwords/blob/d23b853a850ceec358a4137d5e520b067ddb7abc/src/lib.rs#L23
```

Comment:
In procfile-cnb, we used `.expect()` for this case rather than `unwrap()`:
https://github.com/heroku/procfile-cnb/blob/1a5e67fdf6ac87532a5c1893a24e123d541ec3a2/src/procfile.rs#L38

Reply:
This review comment isn't resolved